core/templates/dev/head/domain/statistics/SuggestionImprovementTaskObjectFactorySpec.ts

```diff
@@ -153,6 +153,26 @@ describe('SuggestionImprovementTaskObjectFactory', function() {
       new LearnerAnswerDetailsObjectFactory());
     $provide.value(
       'LearnerAnswerInfoObjectFactory', new LearnerAnswerInfoObjectFactory());
+    $provide.value(
+      'ThreadDataService', {
+        data: {
+          suggestionThreads: [{
+            threadId: 'abc1'
+          }, {
+            threadId: 'def2'
+          }]
+        },
+        fetchThreads: function() {
```

Comment:
Can we use spyOn to call a mock function? This test file assumes too much about the file it is testing. That should not be the case. This way, if the fetchThreads function is deleted in the code, the test will still continue to pass.

Reply:
Hey @nithusha21, sorry for the delay with this PR. I'm facing some issues trying to get `SuggestionImprovementTaskObjectFactory`.ts and `FeedbackImprovementTaskObjectFactory`.ts to 100% coverage using spyOn().

I've found that `return ThreadDataService.fetchMessages(feedback.threadId)` line from this piece of code is the unique line which is being missed.
```
fetchTasks: function() {
        var createNew = this.createNew;
        return ThreadDataService.fetchThreads().then(function() {
          return $q.all(
            ThreadDataService.data.feedbackThreads.map(function(feedback) {
              return ThreadDataService.fetchMessages(feedback.threadId);
            }));
        }).then(function() {
          return ThreadDataService.getData().feedbackThreads.map(createNew);
        });
      },
```

Here's the code I'm trying to do using spyOn():
![Screenshot from 2019-12-13 09-43-11](https://user-images.githubusercontent.com/34922478/70801210-0d5cc800-1d8d-11ea-820b-17b9ecbd8b8f.png)

The both files is coverage by 96.43% and just one line is missing. When I remove the `fetchMessages` method, it comes up to 100%. I don't know why this line is still being missed. Am I doing something wrong? Do you have any suggestion? Thanks!

Reply:
Hey @nithusha21, I've fixed the services coverage, thanks a lot for the suggestion of using $httpBackend. Actually I didn't use it but it helped me a lot to find the right solution.